core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationModule.java

```diff
@@ -116,7 +116,7 @@ protected void configure() {
 
         // bind ShardsAllocator
         final String shardsAllocatorType = settings.get(AllocationModule.SHARDS_ALLOCATOR_TYPE_KEY, AllocationModule.BALANCED_ALLOCATOR);
-        final Class<? extends ShardsAllocator> shardsAllocator = shardsAllocators.get(shardsAllocatorType);
+        final Class<? extends ShardsAllocator> shardsAllocator = shardsAllocators.getExtension(shardsAllocatorType);
```

Comment:
I wonder if we could also factor out the lookup in settings and selection of the class? So we could have a bind method like below, but which takes the binder, Settings and the setting name? In this case, we can just have a lone if to check for the type and log the deprecation message (a little more work here, but a very unusual case)?